src/backend/graphics/egl.rs

```diff
@@ -497,49 +525,56 @@ impl EGLContext {
             out
         };
 
-        trace!(log, "Creating EGL window surface...");
-        let surface = {
-            let surface = match native {
-                Native::X11(_, window) |
-                Native::Wayland(_, window) |
-                Native::Gbm(_, window) => {
-                    egl.CreateWindowSurface(display, config_id, window, surface_attributes.as_ptr())
-                }
-            };
-
-            if surface.is_null() {
-                return Err(CreationError::OsError(String::from("eglCreateWindowSurface failed")));
-            }
-            surface
-        };
-        debug!(log, "EGL window surface successfully created");
-
         info!(log, "EGL context created");
+
         Ok(EGLContext {
                _lib: lib,
                context: context as *const _,
                display: display as *const _,
                egl: egl,
-               surface: surface as *const _,
+               config_id: config_id,
+               surface_attributes: surface_attributes,
                pixel_format: desc,
+               logger: log,
            })
     }
 
-    /// Swaps buffers at the end of a frame.
-    pub fn swap_buffers(&self) -> Result<(), SwapBuffersError> {
-        let ret = unsafe {
-            self.egl
-                .SwapBuffers(self.display as *const _, self.surface as *const _)
-        };
+    /// Creates a surface bound to the given egl context for rendering
+    ///
+    /// # Unsafety
+    ///
+    /// This method is marked unsafe, because the contents of `NativeSurface` cannot be verified and msy
+    /// contain dangeling pointers are similar unsafe content
+    pub unsafe fn create_surface<'a>(&'a self, native: NativeSurface)
+                                     -> Result<EGLSurface<'a>, CreationError> {
+        trace!(self.logger, "Creating EGL window surface...");
 
-        if ret == 0 {
-            match unsafe { self.egl.GetError() } as u32 {
-                ffi::egl::CONTEXT_LOST => Err(SwapBuffersError::ContextLost),
-                err => panic!("eglSwapBuffers failed (eglGetError returned 0x{:x})", err),
+        let surface = {
+            let surface = match native {
+                NativeSurface::X11(window) |
+                NativeSurface::Wayland(window) |
+                NativeSurface::Gbm(window) => {
```

Comment:
What an oversight, thanks. I will add a run time check (the methods are unsafe anyway and all variants contain the same type, compile time checks through new types don't make a lot of sense).